content/blog/aws-serverless-analytics/index.md

```diff
@@ -0,0 +1,604 @@
+
+Pulumi takes care of the complexity by bringing software development best practices like abstraction, encapsulation, and modularity to infrastructure. You can use design patterns in your favorite programming language to provision modern cloud infrastructure. [Component Resources](https://www.pulumi.com/blog/creating-and-reusing-cloud-components-using-package-managers/) are our tools for encapsulating low-level cloud infrastructure into Architecture as Code that is shareable across teams.
+
+With Pulumi the code we write looks very similar to our architecture diagram:
+
+```typescript
+import { ServerlessDataWarehouse, StreamingInputTableArgs }
+    from "./datawarehouse";
+
+// declare our shared schema for clicks and impressions
+const columns = [
+    {
+        name: "id",
+        type: "string",
+    },
+    {
+        name: "session_id",
+        type: "string",
+    },
+    {
+        name: "event_time",
+        type: "string",
+    },
+];
+
+const impressionsTableName = "impressions";
+const clicksTableName = "clicks";
+
+const tableArgs: StreamingInputTableArgs = {
+    columns,
+    inputStreamShardCount: 1,
+    region: "us-west-2",
+};
+
+// create two tables with kinesis input streams,
+// writing data into hourly partitions in S3.
+const dataWarehouse = new ServerlessDataWarehouse("analytics_dw", { isDev })
+    .withStreamingInputTable(impressionsTableName, genericTableArgs)
+    .withStreamingInputTable(clicksTableName, genericTableArgs);
+
+// Make input streams available to external programs as stack outputs
+export const impressionInputStream = impressionsInputStream.name;
+export const clickInputStream = clicksInputStream.name;
+```
+
+In a dozen lines of code, we’ve provisioned two of our desired tables, “clicks” and “impressions”, using our serverless streaming input architecture. It takes care of everything, including implicitly creating an arrival time [partition scheme](https://docs.aws.amazon.com/athena/latest/ug/partitions.html) with the key “inserted_at”. We export our Kinesis input streams as [stack outputs](/docs/intro/concepts/stack/#outputs) that can be referenced in other projects, such as the instrumentation within our ad server or consumer-facing web app.
+
+While on the surface this Pulumi component is described imperatively, it produces a declarative output in the form of a [state file](/docs/intro/concepts/state/) that can be managed locally, in an object store like S3, or by the Pulumi Service backend. Running a ‘pulumi up’ shows that we’ve created 45 AWS resources, and lists our stack outputs to the console.
```

Comment:
```suggestion
While on the surface this Pulumi component is described imperatively, it produces a declarative output in the form of a [state file]({{<  rel "docs/intro/concepts/statemd" >}}) that can be managed locally, in an object store like S3, or by the Pulumi Service backend. Running a ‘pulumi up’ shows that we’ve created 45 AWS resources, and lists our stack outputs to the console.
```

Hugo uses `ref` and `relref` shortcodes to render internal links

Reply:
👍